compiler/src/dotty/tools/dotc/transform/IsInstanceOfEvaluator.scala

```diff
@@ -128,6 +134,67 @@ class IsInstanceOfEvaluator extends MiniPhaseTransform { thisTransformer =>
           val selClassNonFinal = selClass && !(selector.typeSymbol is Final)
           val selFinalClass    = selClass && (selector.typeSymbol is Final)
 
+          // Check if the selector's potential type parameters will be erased, and if so warn
+          val selTypeParam = tree.args.head.tpe.widen match {
+            case tp @ AppliedType(_, args @ (arg :: _)) =>
+              // If the type is `Array[X]` where `X` is a primitive value
+              // class. In the future, when we have a solid implementation of
+              // Arrays of value classes, we might be able to relax this check.
+              val anyValArray = tp.isRef(defn.ArrayClass) && arg.typeSymbol.isPrimitiveValueClass
+              // param is: Any | AnyRef | java.lang.Object
+              val topType = defn.ObjectType <:< arg
+              // has @unchecked annotation to suppress warnings
+              val hasUncheckedAnnot = arg.hasAnnotation(defn.UncheckedAnnot)
+
+              // Shouldn't warn when matching on a subclass with underscore
+              // params or type binding
+              val matchingUnderscoresOrTypeBindings = args.forall(_ match {
+                case tr: TypeRef =>
+                  tr.symbol.is(BindDefinedType)
+                case TypeBounds(lo, hi) =>
+                  (lo eq defn.NothingType) && (hi eq defn.AnyType)
```

Comment:
`eq` usually works, but is not in general a reliable method to compare types, since hash consing is not guaranteed. I recommend to use instead:

    lo.isRef(defn.NothingClass) && hi.isRef(defn.AnyClass)